### PR TITLE
fix: paginate space share

### PR DESCRIPTION
**Pull request by almeidabbm**
Closes: [#12006](https://github.com/lightdash/lightdash/issues/12006)

### Description:

To test, apply the following `diff` and re-run `sh ./scripts/seed-lightdash.sh` - this will add 10k users to the database

```diff
diff --git a/packages/backend/src/database/seeds/development/01_initial_user.ts b/packages/backend/src/database/seeds/development/01_initial_user.ts
index 395333e8b..16b97cbc5 100644
--- a/packages/backend/src/database/seeds/development/01_initial_user.ts
+++ b/packages/backend/src/database/seeds/development/01_initial_user.ts
@@ -20,6 +20,7 @@ import {
 import bcrypt from 'bcrypt';
 import { Knex } from 'knex';
 import path from 'path';
+import { v4 } from 'uuid';
 import { lightdashConfig } from '../../../config/lightdashConfig';
 import { ProjectModel } from '../../../models/ProjectModel/ProjectModel';
 import { projectAdapterFromConfig } from '../../../projectAdapters/projectAdapter';
@@ -35,16 +36,27 @@ export async function seed(knex: Knex): Promise<void> {
     await knex('organizations').del();
 
     const addUser = async (
-        seedOrganization: DbOrganizationIn,
+        seedOrganization: DbOrganizationIn | undefined,
         seedUser: DbUserIn,
         seedEmail: Omit<DbEmailIn, 'user_id'>,
         seedPassword: { password: string },
+        existingOrgUuid?: string,
     ) => {
-        const [{ organization_id: organizationId }] = await knex(
-            'organizations',
-        )
-            .insert(seedOrganization)
-            .returning('organization_id');
+        let organizationId: number | undefined;
+
+        if (seedOrganization) {
+            const [{ organization_id: o }] = await knex('organizations')
+                .insert(seedOrganization)
+                .returning('organization_id');
+
+            organizationId = o;
+        } else if (existingOrgUuid) {
+            const [org] = await knex('organizations')
+                .select()
+                .where('organization_uuid', existingOrgUuid);
+            organizationId = org.organization_id;
+        }
+
         if (organizationId === undefined) {
             throw new Error('Organization was not created');
         }
@@ -85,6 +97,34 @@ export async function seed(knex: Knex): Promise<void> {
         SEED_ORG_1_ADMIN_EMAIL,
         SEED_ORG_1_ADMIN_PASSWORD,
     );
+
+    async function addXUsers(x: number) {
+        const promises: ReturnType<typeof addUser>[] = [];
+        // eslint-disable-next-line no-plusplus
+        for (let i = 1; i <= x; ++i) {
+            promises.push(
+                addUser(
+                    undefined,
+                    {
+                        ...SEED_ORG_1_ADMIN,
+                        first_name: `${SEED_ORG_1_ADMIN.first_name} ${i}`,
+                        user_uuid: v4(),
+                    },
+                    {
+                        ...SEED_ORG_1_ADMIN_EMAIL,
+                        email: `demo+${i}@lightdash.com`,
+                    },
+                    SEED_ORG_1_ADMIN_PASSWORD,
+                    SEED_ORG_1.organization_uuid,
+                ),
+            );
+        }
+
+        await Promise.all(promises);
+    }
+
+    await addXUsers(10000);
+
     await addUser(
         SEED_ORG_2,
         SEED_ORG_2_ADMIN,
```

Note: this does not add backend pagination to space groups/share access, see `packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx`

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceAddUser.tsx [hunk 1]
```diff
@@ -215,7 +251,34 @@ export const ShareSpaceAddUser: FC<ShareSpaceAddUserProps> = ({
                 onChange={setUsersSelected}
                 data={data}
                 itemComponent={UserItemComponent}
-                filter={filterByNameOrEmail}
```
Comment (almeidabbm) [inline packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceAddUser.tsx:218]: No need to perform search by email here since the backend now takes care of all that

New filter function just let's all users go (ignores filtering), and matches the groups

File: packages/backend/src/models/GroupsModel.ts [hunk 2]
```diff
@@ -29,27 +37,45 @@ export class GroupsModel {
...
-        const groups = await query;
-        return groups.map((group) => ({
-            uuid: group.group_uuid,
-            name: group.name,
-            createdAt: group.created_at,
-            organizationUuid: group.organization_uuid,
-        }));
+
+        if (filters.searchQuery) {
+            query = getColumnMatchRegexQuery(query, filters.searchQuery, [
```
Comment (IrakliJani) [inline packages/backend/src/models/GroupsModel.ts:60]: I see you've been adding nice DB utilities over the year! 😆

File: packages/backend/src/models/GroupsModel.ts [hunk 3]
```diff
@@ -29,27 +37,45 @@ export class GroupsModel {
...
+                'name',
+            ]);
+        }
+
+        const { pagination, data } = await KnexPaginate.paginate(
```
Comment (IrakliJani) [inline packages/backend/src/models/GroupsModel.ts:65]: ahaha, strengthens my previous statement!

File: packages/frontend/src/hooks/useOrganizationUsers.ts [hunk 4]
```diff
@@ -86,20 +96,64 @@ export const usePaginatedOrganizationUsers = (params: {
...
+        searchInput?: string;
+        includeGroups?: number;
+        pageSize: number;
+    },
+    infinityQueryOpts: UseInfiniteQueryOptions<
+        ApiOrganizationMemberProfiles['results'],
+        ApiError
+    > = {},
+) => {
+    const setErrorResponse = useQueryError();
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/useOrganizationUsers.ts:124]: What does this do?

File: packages/frontend/src/hooks/useOrganizationGroups.ts [hunk 5]
```diff
@@ -1,62 +1,111 @@
...
-    return useQuery<GroupWithMembers[], ApiError>({
-        queryKey: ['organization_groups', includeMembers],
-        queryFn: () => getOrganizationGroupsQuery(includeMembers),
+    return useQuery<ApiGroupListResponse['results']['data'], ApiError>({
+        queryKey: ['organization_groups', includeMembers, searchInput],
+        queryFn: async () => {
+            return (
+                await getOrganizationGroupsQuery(includeMembers, searchInput)
+            ).data;
+        },
```
Comment (IrakliJani) [inline packages/frontend/src/hooks/useOrganizationGroups.ts:64]: this is basically an old hook without pagination, right?

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceAddUser.tsx [hunk 6]
```diff
@@ -215,7 +251,34 @@ export const ShareSpaceAddUser: FC<ShareSpaceAddUserProps> = ({
...
+                                variant="white"
+                                onClick={async () => {
+                                    await Promise.all([
+                                        fetchGroupsNextPage(),
+                                        fetchUsersNextPage(),
+                                    ]);
+                                }}
+                                disabled={isUsersFetching || isGroupsFetching}
+                            >
+                                <Text size="xs">Load more</Text>
```
Comment (IrakliJani) [inline packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceAddUser.tsx:268]: button with small text? maybe a small button instead? I'll check in the UI

File: packages/frontend/src/components/common/PaginateControl.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                <Button
+                    size="xs"
+                    variant="outline"
+                    color="gray.7"
+                    onClick={onNextPage}
+                    disabled={!hasNextPage}
+                >
+                    <MantineIcon icon={IconChevronRight} />
+                </Button>
+            </Button.Group>
```
Comment (IrakliJani) [inline packages/frontend/src/components/common/PaginateControl.tsx:57]: there's this https://v6.mantine.dev/core/pagination/ if you wanna use it here

File: packages/frontend/src/components/UserSettings/UsersAndGroupsPanel/GroupsView.tsx [hunk 8]
```diff
@@ -226,24 +226,29 @@ const GroupsView: FC = () => {
...
-                                        setGroupToEdit(g);
-                                        setShowCreateAndEditModal(true);
-                                    }}
-                                    onDelete={(groupForDeletion) => {
-                                        setGroupToDelete(groupForDeletion);
-                                        setIsDeleteDialogOpen(true);
-                                    }}
-                                />
-                            ))
+                            groups?.map((group) => {
```
Comment (IrakliJani) [inline packages/frontend/src/components/UserSettings/UsersAndGroupsPanel/GroupsView.tsx:229]: `?` is not required here I guess

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx [hunk 9]
```diff
@@ -296,15 +333,43 @@ type GroupAccessListProps = {
...
 };
 const GroupsAccessList: FC<GroupAccessListProps> = ({
     isPrivate,
     onAccessChange,
     groupsAccess,
+    pageSize,
 }) => {
+    const [page, setPage] = useState(1);
+
+    // TODO: Paginate group access from backend
```
Comment (IrakliJani) [inline packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:346]: do we need to do this in the same PR?

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx [hunk 10]
```diff
@@ -296,15 +333,43 @@ type GroupAccessListProps = {
...
+    const paginatedList: SpaceGroup[][] = useMemo(() => {
+        const list: SpaceGroup[][] = [];
+        const accessListClone = cloneDeep(groupsAccess);
```
Comment (IrakliJani) [inline packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:349]: why is this necessary?

Comment (IrakliJani) [reply packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:349]: fyi, if it's absolutely necessary: https://www.builder.io/blog/structured-clone

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx [hunk 11]
```diff
@@ -145,146 +148,180 @@ type UserAccessListProps = {
...
+    pageSize,
 }) => {
+    const [page, setPage] = useState(1);
+
+    // TODO: Paginate space access from backend
+    const paginatedList: SpaceShare[][] = useMemo(() => {
+        const list: SpaceShare[][] = [];
+        const accessListClone = cloneDeep(accessList).sort(
+            sortByRole(sessionUser?.userUuid),
+        );
```
Comment (IrakliJani) [inline packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:167]: same here

File: packages/frontend/src/hooks/useOrganizationUsers.ts (hunk 4 above)
Comment (almeidabbm) [reply packages/frontend/src/hooks/useOrganizationUsers.ts:124]: This is a hook that processes error responses and shows the toast depending on the error 👌

File: packages/frontend/src/hooks/useOrganizationGroups.ts (hunk 5 above)
Comment (almeidabbm) [reply packages/frontend/src/hooks/useOrganizationGroups.ts:64]: Yep, this is the exact same request with no pagination, the difference is the returned type since now the endpoint supports pagination we return the data in `results.data`

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx (hunk 10 above)
Comment (almeidabbm) [reply packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:349]: This is necessary because we're using splice which mutates the array.
Changed for `structuredClone` 👌

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx (hunk 9 above)
Comment (almeidabbm) [reply packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:346]: We need to paginate these because otherwise we will try to render all of the users/groups in the lists, the problem with these is that they come from nested properties in the request responses, so we can't really paginate these requests (there should be endpoints specific to fetch these 2 lists)

We have created tickets which are part of this milestone to do this, but for now a simple frontend pagination helps to render the modal and avoid the sluggish behavior

File: packages/frontend/src/components/common/PaginateControl.tsx (hunk 7 above)
Comment (almeidabbm) [reply packages/frontend/src/components/common/PaginateControl.tsx]: Made this change, this also impacts the table pagination but it works well and it is a bit better

image

File: packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx [hunk 12]
```diff
@@ -145,146 +147,180 @@ type UserAccessListProps = {
...
+        const accessListClone = structuredClone(accessList).sort(
+            sortByRole(sessionUser?.userUuid),
+        );
+
+        while (accessListClone.length) {
+            list.push(accessListClone.splice(0, pageSize ?? DEFAULT_PAGE_SIZE));
+        }
+
+        return list;
+    }, [accessList, pageSize, sessionUser?.userUuid]);
```
[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:173]: I think you can use this https://lodash.com/docs#chunk

[the remark this answers is not in the document]
Comment (IrakliJani) [reply packages/frontend/src/components/common/ShareSpaceModal/ShareSpaceUserList.tsx:173]:
```ts
const paginatedList: SpaceShare[][] = useMemo(() => {
    const sortedList = structuredClone(accessList).sort(
        sortByRole(sessionUser?.userUuid)
    );
    return chunk(sortedList, pageSize ?? DEFAULT_PAGE_SIZE);
}, [accessList, pageSize, sessionUser?.userUuid]);
```

**Comment by IrakliJani** [comment]
A few things on the user's page list:

https://github.com/user-attachments/assets/d650db96-c204-4600-b392-9c0a5faa0093

- The loader is large and pinned to the top. I usually handle this by using `<Center><Loader color="gray" size="sm" /></Center>`.
- Additionally, you can pass the `keepPreviousData` parameter to the query, which will prevent the existing page content from disappearing. You could also use the `LoadingOverlay` component to indicate the loading state.

Space list:

https://github.com/user-attachments/assets/a82be27d-09dd-44b8-b3b6-db8a9dcd20d6

- There's no loading state, so it's difficult to determine what happened.

**Review by IrakliJani — APPROVED** [review]
approving code changes 👍 great work!
